compiler/src/dotty/tools/dotc/ast/Trees.scala

```diff
@@ -855,16 +855,30 @@ object Trees {
    *                            if this is of class untpd.DerivingTemplate.
    *                            Typed templates only have parents.
    */
-  case class Template[+T <: Untyped] private[ast] (constr: DefDef[T], parentsOrDerived: List[Tree[T]], self: ValDef[T], private var preBody: LazyTreeList[T])(implicit @constructorOnly src: SourceFile)
+  case class Template[+T <: Untyped] private[ast] (constr: DefDef[T], private var myParentsOrDerived: LazyTreeList[T], self: ValDef[T], private var preBody: LazyTreeList[T])(implicit @constructorOnly src: SourceFile)
     extends DefTree[T] with WithLazyField[List[Tree[T]]] {
     type ThisTree[+T <: Untyped] = Template[T]
     def unforcedBody: LazyTreeList[T] = unforced
     def unforced: LazyTreeList[T] = preBody
+
     protected def force(x: List[Tree[T @uncheckedVariance]]): Unit = preBody = x
```

Comment:
These methods from `WithLazyField` won't cause `myParentsOrDerived` to be forced which may break various tree traversals where we call `WithLazyField#forceIfLazy`. It seems that `WithLazyField` needs to be generalized to support multiple lazy fields.

Reply:
In fact they do force `myParentsOrDerived` by the logic in `Template` but that logic is obscure. I cleaned it up.